src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,295 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static SyntaxFactory;
+
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        [Flags]
+        private enum NodeKind
+        {
+            // Represents a combination of either
+            //      two operands of the &&-operator.
+            //      two patterns against the same expression.
+            Conjunction = 1,
+
+            // Represents the var-pattern
+            VarPattern = 1 << 1,
+
+            // Represents a type-check either via an as-expression or a direct cast.
+            TypePattern = 1 << 2,
+
+            // Represents a negative null comparison.
+            NotNullPattern = 1 << 3,
+
+            // Represents a constant pattern or expression.
+            ConstantPattern = 1 << 4,
+
+            // Represents an equality comparison or an is-expression.
+            PatternMatch = 1 << 5,
```

Comment:
why not call it EqualityComparisonOrIsExpression?  `PatternMatch` does not at all evoke that concept to me.

Reply:
because it won't remain an EqualityComparisonOrIsExpression throughout the transformation. this is solely to represent a pattern match against some expression, it can come from an is-expression, equality operator or even a simple identifier, at the end we decide that it should be an is-expression or a subpattern! maybe you'd prefer IsPattern?

Reply:
Basically, i need you to expand hte docs here such that the name makes sense.  Perhaps that means showing what you start with and what it would turn into.  Then maybe hte name will be clearer :)